### PR TITLE
Aggregated Search

### PR DESCRIPTION
**Pull request by bill-filler**
This PR aggregates search across pages, documents and videos.
Major changes are the following: sorry if this is too big. I can split

1. New model: BookDocument
2. Modified model: Video - no longer store transcript, but instead transcript_url. Transcript will be fetched as part of search indexing via the def transcript() method on Video
3. Modified search backend to support proper indexing of document contents via ingest-attachment, and optimize settings for large text fields needed by video and documents
4. Modified search/views.py to query across multiple indices and aggregate the results for display
5. Added a utils.py: to be able to find page that references a particular video/document as there is no specific foreign-key relationship defined in models (NOTE: we should consider adding such a relationship, as this method likely will not scale well)
6. Created a very rough JwtLmsApiClient that now houses the code that used to live in the mgmt command such that it can be used from multiple places.

### CODE REVIEW

File: digitalbook/book/utils.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+from .models import BookPage
+import blocks
+
+
+def get_page_for_object(book_pages, object_pk, object_type):
```
Comment (tuchfarber) [inline digitalbook/book/utils.py]: Based on our conversation, this is going away. That makes me happy.

File: digitalbook/search/views.py [hunk 2]
```diff
@@ -1,36 +1,108 @@
...
+        #import pdb; pdb.set_trace()
+
+        sorted_results = sorted(results.values(), key=lambda page: page.max_score, reverse=True)
+
+        # Record hit, this is for promotions
+        query = Query.get(search_query)
         query.add_hit()
     else:
-        search_results = Page.objects.none()
+        page_search_results = Page.objects.none()
```
Comment (tuchfarber) [inline digitalbook/search/views.py]: These 3 vars are unused.

File: digitalbook/search/views.py [hunk 3]
```diff
@@ -1,36 +1,108 @@
...
+    #     page_search_results = paginator.page(page)
+    # except PageNotAnInteger:
+    #     page_search_results = paginator.page(1)
+    # except EmptyPage:
+    #     page_search_results = paginator.page(paginator.num_pages)
 
     return render(request, 'search/search.html', {
         'search_query': search_query,
-        'search_results': search_results,
+        'search_results': sorted_results
```
Comment (tuchfarber) [inline digitalbook/search/views.py]: sorted_results will be undefined if you don't have a search_query

File: digitalbook/book/search/backend.py [hunk 4]
```diff
@@ -1,28 +1,173 @@
...
+                    }
+
+            results = self.es.index(
+                index=INGEST_ATTACHMENT_INDEX,
+                doc_type=INGEST_ATTACHMENT_DOC_TYPE,
+                id=INGEST_ATTACHMENT_ID,
+                body=body
+            )
+            #import pdb;pdb.set_trace()
+            print('created index for _ingest attachment, results=', results)
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/search/backend.py]: fixed

File: digitalbook/book/search/backend.py [hunk 5]
```diff
@@ -1,28 +1,173 @@
...
+
+    def add_item(self, item):
+        '''
+        Called when new item added to the index
+        Need to override to include pipeline attribute for ingest_plugin attachment
+        '''
+        #import pdb;pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/search/backend.py]: fixed

File: digitalbook/book/search/backend.py [hunk 6]
```diff
@@ -1,28 +1,173 @@
...
+        mapping = self.mapping_class(item.__class__)
+        if mapping.get_document_type() == BOOK_DOCUMENT_TYPE:
+            results = self.es.index(
+                self.name,
+                mapping.get_document_type(),
+                mapping.get_document(item),
+                pipeline=INGEST_ATTACHMENT_ID,
+                id=mapping.get_document_id(item)
+            )
+            print('in add_item with attachment results=', results)
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/search/backend.py]: fixed

File: digitalbook/book/search/backend.py [hunk 7]
```diff
@@ -1,28 +1,173 @@
...
+                action = {
+                    '_index': self.name,
+                    '_type': doc_type,
+                    '_id': mapping.get_document_id(item),
+                    'pipeline': INGEST_ATTACHMENT_ID,
+                }
+                action.update(mapping.get_document(item))
+                actions.append(action)
+
+            print('in add_items with attachment')
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/search/backend.py]: fixed

File: digitalbook/book/search/backend.py [hunk 8]
```diff
@@ -31,18 +176,30 @@ def _do_search(self):
...
+            'post_tags': ['</b>']
+        }
         params[self.fields_param_name] = 'pk'
 
         # Add size if set
         if self.stop is not None:
             params['size'] = self.stop - self.start
 
         # Send to Elasticsearch
+        import pprint as pp
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/search/backend.py]: fixed

File: digitalbook/book/utils.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    '''
+    Find the page(s) that reference the specified object
+    Args:
+        object_pk: the primary key of the related object
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/utils.py]: fixed

File: digitalbook/book/utils.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            if object_type == type:
+                # TODO: search for images/docs/videos embedded in RichText block as well
+                if object_type == blocks.PDF_BLOCK_TYPE and object_pk == data.get('value').get(blocks.STREAM_DATA_DOC_FIELD) or \
+                        object_type == blocks.VIDEO_BLOCK_TYPE or blocks.IMAGE_BLOCK_TYPE and object_pk == data.get('value'):
+                            page_set.add(page)
+
+    return page_set
+
+
+def get_related_objects_from_page(book_pages, documents=True, videos=True, images=True):
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/book/utils.py]: removed

File: digitalbook/search/templates/search/search.html [hunk 11]
```diff
@@ -7,36 +7,49 @@
...
-                            {{ result.search_description }}
-                        {% endif %}
-                        {% if result.search_results_metadata %}
-                            {% for key, value in result.search_results_metadata.items %} 
-                                {% for val in value %}
-                                <div id="content-highlight">
-                                    <p>{{val|safe}}</p>
-                                </div>
+                        <h4><a href="{% pageurl search_display.book_page %}">{{ search_display.book_page }}</a>
+                        {% comment %}
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/search/templates/search/search.html]: it comments out the lines until endcomment tag. I removed them though

File: digitalbook/search/templates/search/search.html [hunk 12]
```diff
@@ -7,36 +7,49 @@
...
+                                </ul>
+                                <br>
+                            {% endif %}
+                        {% endfor %}
                     </li>
                 {% endfor %}
             </ul>
-
+            
+            {% comment %}
```
[the remark this answers is not in the document]
Comment (bill-filler) [reply digitalbook/search/templates/search/search.html]: removed

File: digitalbook/book/utils.py (hunk 1 above)
Comment (bill-filler) [reply digitalbook/book/utils.py]: yes is going away

File: digitalbook/search/views.py (hunk 2 above)
Comment (bill-filler) [reply digitalbook/search/views.py]: removed

File: digitalbook/search/views.py (hunk 3 above)
Comment (bill-filler) [reply digitalbook/search/views.py]: fixed

**Pull request merged**